lib/lrama/grammar.rb

```diff
@@ -306,6 +306,48 @@ def nterms
       @nterms ||= @symbols.select(&:nterm?)
     end
 
+    def extract_references
+      @_rules.each do |lhs, rhs, _|
+        rhs.each_with_index do |token, index|
+          next if token.type != Lrama::Lexer::Token::User_code
+
+          scanner = StringScanner.new(token.s_value)
+          references = []
+
+          while !scanner.eos? do
+            start = scanner.pos
+            case
+            when scanner.scan(/\$(<[a-zA-Z0-9_]+>)?\$/) # $$, $<long>$
```

Comment:
If don't need to use the result of `scan`, you can try to use `skip` and save a lot of allocations and improve the performance

Reply:
Similar what I did in this PR https://github.com/ruby/lrama/pull/102